lib/Dialect/Gluon/Transforms/ResolveAutoEncodings.cpp

```diff
@@ -25,222 +18,31 @@ namespace mlir::triton::gluon {
-    if (auto opResult = dyn_cast<OpResult>(val)) {
-      auto definingOp = opResult.getOwner();
-      if (isa<scf::ForOp, scf::WhileOp, scf::IfOp>(definingOp)) {
-        auto tiedArgs = getTiedArgs(definingOp, opResult.getResultNumber());
-        if (failed(updateEncoding(tiedArgs, info)))
-          return failure();
-      } else {
-        auto srcEncoding = inferSrcEncoding(definingOp, info.encoding);
-        if (srcEncoding) {
-          bool mayVary = info.mayVary || encodingsMayVary(definingOp);
-          LayoutInfo srcInfo{srcEncoding, mayVary};
-          llvm::SmallVector<Value> tensorOperands;
-          for (auto operand : definingOp->getOperands())
-            if (isa<RankedTensorType>(operand.getType()))
-              tensorOperands.push_back(operand);
-
-          if (failed(updateEncoding(tensorOperands, srcInfo)))
-            return failure();
-        }
-      }
-    } else if (auto blockArg = dyn_cast<BlockArgument>(val)) {
-      auto parentOp = blockArg.getOwner()->getParentOp();
-      if (isa<scf::ForOp, scf::WhileOp>(parentOp)) {
-        auto offset = isa<scf::ForOp>(parentOp);
-        auto tiedArgs = getTiedArgs(parentOp, blockArg.getArgNumber() - offset);
-        if (failed(updateEncoding(tiedArgs, info)))
-          return failure();
-      }
-    }
-  }
-
-  // 3. Transfer propagated encodings into the graph
-  auto ctx = func.getContext();
-  for (auto &[val, info] : valueToEncoding) {
-    auto existingTy = cast<RankedTensorType>(val.getType());
-    assert(isa<gluon::AutoEncodingAttr>(existingTy.getEncoding()));
-    auto ty = existingTy.cloneWithEncoding(info.encoding);
-    val.setType(ty);
-
-    if (auto opResult = dyn_cast<OpResult>(val)) {
-      if (auto constantOp = dyn_cast<arith::ConstantOp>(opResult.getOwner())) {
-        auto value = cast<SplatElementsAttr>(constantOp.getValueAttr());
-        auto newValue =
-            SplatElementsAttr::get(ty, value.getSplatValue<Attribute>());
-        constantOp.setValueAttr(newValue);
-      }
-    }
-  }
-
-  // 4. Cleanup set_auto_layout ops
-  func.walk([&](gluon::SetAutoLayoutOp op) {
-    assert(op.getSrc().getType() == op.getType());
-    op.getResult().replaceAllUsesWith(op.getSrc());
-    op->erase();
-  });
-
-  return success();
-}
-
-LogicalResult inferAutoLayouts(ModuleOp &mod) {
+LogicalResult inferAutoLayout(ModuleOp &mod,
+                              llvm::function_ref<bool(Type)> typeCheck) {
   for (auto &op : *mod.getBody()) {
     auto func = dyn_cast<FuncOp>(&op);
     if (!func)
       continue;
-    if (failed(inferAutoLayouts(func)))
+
+    // Set seed values from set_auto_layout ops
+    llvm::SmallVector<std::pair<Value, Attribute>> seedEncodings;
+    auto res = func.walk([&](gluon::SetAutoLayoutOp op) -> WalkResult {
+      seedEncodings.push_back({op.getSrc(), op.getType().getEncoding()});
+      return WalkResult::advance();
+    });
+    if (res.wasInterrupted())
+      return failure();
```

Comment:
You never return `WalkResult::interrupt` so this will never happen.
```suggestion
    func.walk([&](gluon::SetAutoLayoutOp op) {
      seedEncodings.push_back({op.getSrc(), op.getType().getEncoding()});
    });
```